### PR TITLE
Data engine

### PR DESCRIPTION
**Pull request by zxrys**
Add data engine to enable users to build their own DPO dataset.

### CODE REVIEW

**Comment by zxrys** [comment]
@yiranyyu @Haoye17

File: data_engine/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Data Engine
+
+## Welcome
+Thank you for using Data Engine.
```
Comment (yiranyyu) [inline data_engine/README.md:4]: 可以省略

File: data_engine/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Data Engine
+
+## Welcome
+Thank you for using Data Engine.  
+This part of the code is used to build the DPO dataset, which you can use for direct training.
```
Comment (yiranyyu) [inline data_engine/README.md:5]: You can specify reward model and instruction model you want to use for generating preference training dataset. The current supported list of reward models and instruction models are listed at ....

File: data_engine/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+## Welcome
+Thank you for using Data Engine.  
+This part of the code is used to build the DPO dataset, which you can use for direct training.  
+You only need to input the reward model (the model trained with DPO which is used for guidance), instruction model (the model you want to train), and your dataset, and we will generate the DPO dataset for you. All you need to do is run the `run_engine.sh` script.
+
+## Usage
+Please refer to the `run_engine.sh` script.
+
+You will need to provide the path and name for both the reward model and the instruction model. Currently, we support the following models: llava-1.5-7b, RLAIF-V-7B, OmniLMM-12B, and RLAIF-V-12B. We are considering adding more models in the future. \
```
Comment (yiranyyu) [inline data_engine/README.md:11]: 按照奖励模型和指令模型分开介绍，分别输出奖励值和被打分的模型输出。这部分不要涉及训练。

File: data_engine/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+If the model you wish to use is not listed, you may need to implement the corresponding code yourself (for model loading, add code to `RLAIF-V/builder`; for answer sampling, refer to `RLAIF-V/llava/llava15_sample_data.py` to see how data is formatted (don't forget to pass `raw_images`) and add call it in `RLAIF-V/data_engine/answer_sampler.py`; for log probability calculation, change data formatting part in `RLAIF-V/data_engine/logps_calculator.py` and `get_multimodal_sample_logps` function in `RLAIF-V/muffin/eval/muffin_inference_logp.py`).
```
Comment (yiranyyu) [inline data_engine/README.md:12]: 这里拆成几个不同的 subsection.

## Generate Rollouts

## Reward collection

## Customize your reward model

## Customize your instruction model

File: data_engine/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+1. `idx`: A unique index for each data entry (this can be a string).
+2. `question`: The question related to the image.
+3. `image`: The column should follow this structure:
+   - `{'bytes': ..., 'path':...}`
+   - `bytes` should be in binary format.
+   - `path` is not strictly required, but to avoid errors, it's better to keep this field (you can set it as an empty string).
+
+You can specify a `--work_dir` to store intermediate files and the final output under this directory (which will actually be a subdirectory within it).
+
+If you encounter errors during generation, you can pass the stage next to the stage that has been completed using the `--continue_from_stage` parameter (0, 1, or 2). When the value is 0, it will start from scratch. (For example, if you've completed stages 0 and 1 but encounter an error during stage 2, you can fix the issue and set `--continue_from_stage 2` to continue from that point.) You can check the `data_engine.py` file for details on what each stage does.
```
Comment (yiranyyu) [inline data_engine/README.md:26]: 就分成三个脚本吧，尽量不要有含糊的信息。

File: data_engine/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+Run:
+```shell
+sh data_engine/run_data_engine.sh
+```
+
+## Conclusion
```
Comment (yiranyyu) [inline data_engine/README.md:33]: 可以看下其他项目比如 MiniCPM-V 的 readme，这里不需要一个 conclusion

File: script/train/llava15_train.sh [hunk 7]
```diff
@@ -1,12 +1,13 @@
...
+export CUDA_VISIBLE_DEVICES=0,1,2,3
 
 task_name=llava15_7b_DPO
 exp_name=llava15_rlaifv
 
 deepspeed ./muffin/train/train_llava15.py \
     --deepspeed ./script/zero2.json  \
-    --model_name_or_path liuhaotian/llava-v1.5-7b \
-    --data_dir ./RLAIF-V-Dataset_logps/ \
+    --model_name_or_path /data/yaoshu/models/llava-v1.5-7b \
```
Comment (yiranyyu) [inline script/train/llava15_train.sh:9]: 这俩参数改到上面接收位置参数或者关键词参数吧，现在默认可以直接跑，但是会报错也有点不友好

File: pyproject.toml
Comment (yiranyyu) [inline pyproject.toml:1]: 这些版本都需要改吗，会不会影响其他结果可复现性

File: omnilmm/train/train_utils.py
Comment (yiranyyu) [inline omnilmm/train/train_utils.py:1]: 这个为什么需要去掉

Comment (yiranyyu) [reply omnilmm/train/train_utils.py:1]: 看起来加到其他脚本了，再检查下是否会影响其他代码的运行。

File: omnilmm/model/omnilmm.py
Comment (yiranyyu) [inline omnilmm/model/omnilmm.py:1]: 改动很多，可以这次先不修改格式。

File: muffin/eval/muffin_inference_logp.py [hunk 8]
```diff
@@ -323,7 +301,7 @@ def inference_logp(model, tokenizer, hf_data, cache_file, image_token_len, img_p
     dataloader = torch_data.DataLoader(dataset, batch_size=1, collate_fn=collate_fn,
                                        num_workers=5, shuffle=False, sampler=InferenceSampler(len(dataset)))
 
-    outputs = get_multimodal_sample_logps(model, dataloader, tokenizer, is_llava15=is_llava15) # win_logp_list, win_avg_logp_list, win_per_token_logp_list, rej_logp_list, rej_avg_logp_list, rej_per_token_logp_list
```
Comment (yiranyyu) [inline muffin/eval/muffin_inference_logp.py:326]: 这个是无效参数吗，是不是有些 caller 代码还会传这个，是否会被影响？

File: muffin/data/datasets.py [hunk 9]
```diff
@@ -18,38 +18,60 @@
...
 class RLAIFVDataset(torch_data.Dataset):
     def __init__(self, data_dir: str, reference_model=None,
                  tokenizer=None, image_token_len=None, img_processor=None, use_im_start_end=True, is_llava15=False):
         super().__init__()
 
         if not op.exists(data_dir):
             os.makedirs(data_dir, exist_ok=True)
-
-        data_path = [file for file in os.listdir(data_dir) if file.endswith('.parquet') and 'logp' in file]
+        if not op.exists(op.join(data_dir, "logps")):
```
Comment (yiranyyu) [inline muffin/data/datasets.py:38]: 这里的改动是什么？

**Review by yiranyyu — CHANGES_REQUESTED** [review]
Great work! This PR supports automatically generating high-quality preference learning dataset efficiently with RLAIF-V models or other reward models and instruction models.

Still, some of the modification should be further revised before permitted to be merged.

File: muffin/eval/muffin_inference_logp.py (hunk 8 above)
Comment (zxrys) [reply muffin/eval/muffin_inference_logp.py:326]: get_multimodal_sample_logps确认不会用到tokenizer，所有调用处已去除tokenizer实参

File: data_engine/README.md (hunk 1 above)
Comment (zxrys) [reply data_engine/README.md:4]: 好的

File: data_engine/README.md (hunk 6 above)
Comment (zxrys) [reply data_engine/README.md:33]: 好的

File: muffin/data/datasets.py (hunk 9 above)
Comment (zxrys) [reply muffin/data/datasets.py:38]: 这里原来的代码有点问题：如果不存在logps则只能去huggingface下载rlaifv数据集，无法读取用户给定的数据集。我把这里改了下，若是存在数据集文件则对文件计算logps，而后自动读取；若是存在logps则和之前的逻辑类似。

File: omnilmm/model/omnilmm.py
Comment (zxrys) [reply omnilmm/model/omnilmm.py:1]: 嘶，这个格式是pycharm自动format的，需要revert吗，似乎不是太好弄

File: pyproject.toml
Comment (zxrys) [reply pyproject.toml:1]: 我印象中transformers版本为4.35时training报错，印象中是说Bfloat不支持还是啥的，升级到4.37后解决了。其他几个基本上都是因为transformers升级为避免版本依赖冲突而一起升级到。确实可能会影响可复现性，可能需要讨论下

File: script/train/llava15_train.sh (hunk 7 above)
Comment (zxrys) [reply script/train/llava15_train.sh:9]: 哦这我的问题，忘删了，这就改

File: omnilmm/train/train_utils.py
Comment (zxrys) [reply omnilmm/train/train_utils.py:1]: 这个改的时候检查过了，其他指向此处的都改成指向constant文件了，应该不会有问题

File: data_engine/README.md (hunk 3 above)
Comment (zxrys) [reply data_engine/README.md:11]: 好的

File: data_engine/README.md (hunk 2 above)
Comment (zxrys) [reply data_engine/README.md:5]: 好的

File: data_engine/README.md [hunk 10]
```diff
@@ -1,36 +1,133 @@
...
+---
+
+### DPO Reward Pipeline
 
-## Conclusion
-If you run into any issues, feel free to contact us by submitting an Issue.
+#### Process Method
 
-Thank you for choosing RLAIF-V. Best wishes for your project!
+Generates rewards using the DPO framework to rank answers. Higher-ranked answers are marked as "chosen," while
```
Comment (yiranyyu) [inline data_engine/README.md:78]: Use RLAIF-V self-feedback guidance with DPO-trained models.

File: data_engine/README.md [hunk 11]
```diff
@@ -1,36 +1,133 @@
...
-If you encounter errors during generation, you can pass the stage next to the stage that has been completed using the `--continue_from_stage` parameter (0, 1, or 2). When the value is 0, it will start from scratch. (For example, if you've completed stages 0 and 1 but encounter an error during stage 2, you can fix the issue and set `--continue_from_stage 2` to continue from that point.) You can check the `data_engine.py` file for details on what each stage does.
+### Divide and Conquer Pipeline
 
-Run:
-```shell
-sh data_engine/run_data_engine.sh
-```
+#### Process Method
+
+Detailed in the corresponding research paper.
```
Comment (yiranyyu) [inline data_engine/README.md:27]: Use RLAIF-V divide-and-conquer strategy to collect AI feedback.

File: data_engine/README_zh.md [hunk 12]
```diff
@@ -1,36 +1,128 @@
...
-```shell
-sh data_engine/run_data_engine.sh
-```
+---
+
+### Divide and Conquer Pipeline
+
+#### 处理方法
+
+具体流程详见论文。
```
Comment (yiranyyu) [inline data_engine/README_zh.md:25]: 中文参考英文修改

**Review by yiranyyu — CHANGES_REQUESTED** [review]
Last step, refine the readme to improve the readability.

File: data_engine/README_zh.md [hunk 13]
```diff
@@ -72,7 +72,7 @@
 
 #### 处理方法
 
-使用 DPO 框架生成奖励以对答案排序。高分答案标记为Chosen，低分答案标记为Rejected。
+将 RLAIF-V 自反馈指导与 DPO 训练模型结合使用。
```
Comment (yiranyyu) [inline data_engine/README_zh.md:75]: 使用 RLAIF-V 基于 DPO-aligned 模型构造的自反馈信号。

**Review by yiranyyu — CHANGES_REQUESTED** [review]
